Alabou review of latest IS-11 specification

alabou opened the pull request:
- Trying to enhance the normative text about intersection. The key aspects are 1) Controller "chooses" a number of Receivers, 2) "such intersection" could be "satisfying" all or some of the Receivers, 3) "such intersection" should represent the "largest consensus" among the Receivers, 4) the user may set preferences about how the largest consensus is obtained.

- Controller must track Sender during IS-05 connection but also after as the signal may change.

- The Receiver state is not directly associated to the active/inactive state of the Receiver (as it is for the Sender).  The Receiver state is about the "active Flow" which is a way of saying about the "active transport file content".

-  Make the sentence "Inputs/Outputs MAY be associated with multiple Senders/Receivers" more general to allow any possible cases.

alabou committed:
Update Overview.md

alabou committed:
Update Data Model.md

alabou committed:
Update Behaviour - Server Side.md

alabou committed:
Update Behaviour - Client Side.md

alabou committed:
Update Behaviour - Client Side.md

alabou committed:
Update Behaviour - Client Side.md

garethsb commented on docs/Data Model.md:
```diff
@@ -10,7 +10,7 @@ In order to ensure consistent behaviour in a facility employing this identity, s
 The IS-04 hierarchy is preserved. Additionally, Senders managed with IS-11 SHOULD be associated with 1 or more Input resources and Receivers managed with IS-11 SHOULD be associated with 1 or more Output resources.
 
-Inputs/Outputs carrying multiple media formats at a time (e.g. video and audio) MAY be associated with multiple Senders/Receivers of given formats.
+Inputs/Outputs MAY be associated with multiple Senders/Receivers, E.g. An Input carrying multiple media formats at a time (e.g. video and audio) MAY be associated with multiple Senders of varying formats, an Input carrying a given media format that is segmented into multiple parts MAY be associated with multiple Senders of the same formats.
```
Should we also add example(s) for many Receivers -> one Output, and if appropriate, one Receiver -> many Outputs? Maybe break into two paras, one for Inputs/Senders and one for Receivers/Outputs?

alabou replied:
Done, split in two paragraphs.

garethsb commented on docs/Behaviour - Client Side.md:
```diff
@@ -2,9 +2,9 @@
 ## Active Constraints of Sender
 
-Before making an [IS-05][IS-05] connection, NMOS Controller chooses a Sender and Receivers. Then NMOS Controller MUST `GET /constraints/supported` from the Sender, MUST collect Receiver Capabilities from the Receivers and make such intersection of them that they can be utilized for building Active Constraints satisfying all the Receivers and using Parameter Constraint URNs supported by the Sender ...
+Before making an [IS-05][IS-05] connection, NMOS Controller chooses a Sender and a number of  Receivers. Then NMOS Controller MUST `GET /constraints/supported` from the Sender, MUST collect Receiver Capabilities from the Receivers and make such intersection of them that they can be utilized for building Active Constraints satisfying all (default) or some (user preferences) of the Receivers and us ...
```
```suggestion
Before making an [IS-05][IS-05] connection, NMOS Controller chooses a Sender and a number of Receivers. Then NMOS Controller MUST `GET /constraints/supported` from the Sender, MUST collect Receiver Capabilities from the Receivers and make such intersection of them that they can be utilized for building Active Constraints satisfying all (by default) or some (depending on user preferences) of the Receivers and using Parameter Constraint URNs supported by the Sender. If the Sender supports fewer Parameter Constraint URNs than used in the Receiver Capabilities, the NMOS Controller SHOULD inform the user about it. After that NMOS Controller MUST `PUT /constraints/active` to the Sender and SHOULD make the IS-05 connections if the Constraints have been applied succesfully to the Sender. After breaking these connections via IS-05, NMOS Controller is RECOMMENDED to `DELETE /constraints/active` of the Sender.
```

garethsb replied:
How about the case where there intersection is empty, where it's impossible to satisfy a Constraint Set of all of the Receivers at once? I assume the Controller SHOULD/MUST inform the user in this case too?

N-Nagorny replied:
Language candidate from the call:

If the processing Constraint Sets results in an empty `constraint_sets` of Active Constraints, NMOS Controller SHOULD/MAY inform the user about it.

garethsb commented on docs/Behaviour - Client Side.md:
```diff
@@ -2,9 +2,9 @@
+Before making an [IS-05][IS-05] connection, NMOS Controller chooses a Sender and a number of  Receivers. Then NMOS Controller MUST `GET /constraints/supported` from the Sender, MUST collect Receiver Capabilities from the Receivers and make such intersection of them that they can be utilized for building Active Constraints satisfying all (default) or some (user preferences) of the Receivers and us ...
 
-Constraint Sets of Active Constraints SHOULD be the largest consensus among Receivers. For example, there are Receiver A, Receiver B, Receiver C with equal Receiver Capabilities consisting of Constraint Sets 1, 2, 3, 4, 5 and Receiver D with Constraint Sets 2, 3, 4, 5 and 6 which is incompatible with the previous ones. In this case the largest consensus is Constraint Sets 2, 3, 4, 5 which are com ...
+Constraint Sets of Active Constraints SHOULD represent the largest consensus among Receivers. For example, there are Receiver A, Receiver B, Receiver C with equal Receiver Capabilities consisting of Constraint Sets 1, 2, 3, 4, 5 and Receiver D with Constraint Sets 2, 3, 4, 5 and 6 which is incompatible with the previous ones. In this case the largest consensus is Constraint Sets 2, 3, 4, 5 which  ...
```
```suggestion
Constraint Sets of Active Constraints SHOULD represent a consensus among Receivers. For example, there are Receiver A, Receiver B, Receiver C with equal Receiver Capabilities consisting of Constraint Sets 1, 2, 3, 4, 5 and Receiver D with Constraint Sets 2, 3, 4, 5 and 6 which is incompatible with the previous ones. In this case the largest consensus is Constraint Sets 2, 3, 4, 5 which are common for all the four Receivers. The user may control how the consensus is obtained, providing preferences to the NMOS Controller among the Receivers and the Constraint Sets of each Receiver and criteria about the size of the consensus.
```

I think Alain's wording is good, I'm just proposing deleting "largest" in two places to allow for the user overriding. Could instead replace with "best" or "preferred" or something...

Perhaps we should use "consensus" throughout this documentation instead of "intersection", apart from as an example of default behaviour when there is no user control? Intersection is almost too precise a term I feel.

garethsb replied:
I think we also need to update the second part of this example two paragraphs lower down, to explain precisely what "more important" means to the user in this example.

alabou replied:
What about replacing "such intersection" by  "such processing" as follow: "Then NMOS Controller MUST `GET /constraints/supported` from the Sender, MUST collect Receiver Capabilities from the Receivers and make such processing of them that they can be utilized for building Active Constraints satisfying all (by default) or some (depending on user preferences) of the Receivers and using Parameter Constraint URNs supported by the Sender."

The idea is the the controller in its processing of the receiver capabilities may perform intersections based on some criteria and may also perform selections based on some other criteria. It is true that the word "intersection" may be misleading ...

alabou committed:
Update docs/Behaviour - Client Side.md

Co-authored-by: Gareth Sylvester-Bradley <31761158+garethsb@users.noreply.github.com>

alabou committed:
Update Behaviour - Client Side.md

remove largest and keep a consensus

alabou committed:
Update Data Model.md

alabou committed:
Update Behaviour - Client Side.md

alabou commented:
I answered to Gareth's comments in 4 additional commits. Make sure to compare using all the commits of alabou-patch-1 from v1.0-dev.

I included the modification to replace "intersection" with "processing".

N-Nagorny commented on docs/Behaviour - Client Side.md:
```diff
@@ -2,22 +2,22 @@
+Constraint Sets of Active Constraints SHOULD represent a consensus among Receivers. For example, there are Receiver A, Receiver B, Receiver C with equal Receiver Capabilities consisting of Constraint Sets 1, 2, 3, 4, 5 and Receiver D with Constraint Sets 2, 3, 4, 5 and 6 which is incompatible with the previous ones. In this case the consensus is Constraint Sets 2, 3, 4, 5 which are common for all ...
 
-`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Receiver Capabilities indicates relative preference between Constraint Sets of that Receiver. The NMOS Controller MAY use these values to make a decision about the intersection and determine preference of Constraint Sets of Active Constraints. It SHOULD NOT propagate values of this property as is from Receiver Capabilities to Active  ...
+`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Receiver Capabilities indicates relative preference between Constraint Sets of that Receiver. The NMOS Controller MAY use these values to make a decision about their processing and determine preference of Constraint Sets of Active Constraints. It SHOULD NOT propagate values of this property as is from Receiver Capabilities to Active  ...
```
Would it be better to say "make a decision about the consensus"?

alabou replied:
The issue I try to resolve with something more general like "processing" is that it is not always a consensus that is the result ... We provide an example where the user select Receiver D-6 which is not part of the consensus ... This is why I tried to be more general using the word "processing" .... TO DISSUSS.

N-Nagorny replied:
```suggestion
`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Receiver Capabilities indicates relative preference between Constraint Sets of that Receiver. The NMOS Controller MAY use these values to make a decision about how to process these Constraint Sets and determine preference of Constraint Sets of Active Constraints. It SHOULD NOT propagate values of this property as is from Receiver Capabilities to Active Constraints.
```

N-Nagorny commented on docs/Behaviour - Client Side.md:
```diff
@@ -2,22 +2,22 @@
+`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Receiver Capabilities indicates relative preference between Constraint Sets of that Receiver. The NMOS Controller MAY use these values to make a decision about their processing and determine preference of Constraint Sets of Active Constraints. It SHOULD NOT propagate values of this property as is from Receiver Capabilities to Active  ...
 
-`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Active Constraints indicates relative preference between Constraints Sets of the Active Constraints of the Sender. The NMOS Controller MAY fill in this property in Constraint Sets based on any additional information about Receivers. For example, if the user in any way informs the NMOS Controller that Receiver D is more important than ...
+`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Active Constraints indicates relative preference between Constraints Sets of the Active Constraints of the Sender. The NMOS Controller MAY fill in this property in Constraint Sets based on any additional information about Receivers. For example, if the user in any way informs the NMOS Controller that Receiver D prefered Constraint Se ...
```
```suggestion
`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Active Constraints indicates relative preference between Constraint Sets of the Active Constraints of the Sender. The NMOS Controller MAY fill in this property in Constraint Sets based on any additional information about Receivers. For example, if the user in any way informs the NMOS Controller that Receiver D preferred Constraint Set has precedence over the consensus of Receivers A, B, C and `urn:x-nmos:cap:meta:preference` values of Receiver D state that Constraint Set 6 has higher preference than Constraints Sets 2, 3, 4, 5, then the Constraint Sets would be 2, 3, 4, 5, 6 where Constraint Set 6 has the highest preference.
```

N-Nagorny commented on docs/Behaviour - Server Side.md:
```diff
@@ -21,9 +21,9 @@ A Sender managed with IS-11 has the following states:
-- `unknown` when this Receiver is inactive or active and it is impossible to say whether the Flow is compliant with Receiver Capabilities of this Receiver (e.g. it is activated without a `transport_file`).
-- `compliant_flow` when this Receiver is active and the Flow is compliant with Receiver Capabilities of this Receiver (e.g. there is an active `transport_file` and it does not violate the Receiver Capabilities).
-- `non_compliant_flow` when this Receiver is active and the Flow is non compliant with Receiver Capabilities of this Receiver (e.g. there is an active `transport_file` which violates the Receiver Capabilities). If the Receiver becomes inactive this state is preserved until the next activation.
+- `unknown` when it is not possible to say whether the active Flow is compliant with Receiver Capabilities of this Receiver (e.g. it is activated without a `transport_file`).
+- `compliant_flow` when the active Flow is compliant with Receiver Capabilities of this Receiver (e.g. there is an active `transport_file` and it does not violate the Receiver Capabilities).
+- `non_compliant_flow` when the active Flow is non compliant with Receiver Capabilities of this Receiver (e.g. there is an active `transport_file` which violates the Receiver Capabilities). If the Receiver becomes inactive this state is preserved until the next activation.
```
Considering that "active Flow" is a not defined term, I think the old definitions are more precise.

alabou replied:
I want to make sure we refer to the active transport file .... the Flow deriving from the active transport file ... that I named the active Flow ... I also attempt to separate those definitions form the active/inactive state as mush as possible as it is for the Sender, except for non_compliant_flow ... TO DISCUSS.

N-Nagorny reviewed (commented):
Thank you for the great work, Alain!

alabou committed:
Update docs/Behaviour - Client Side.md from Nikita comment

Co-authored-by: N-Nagorny <Nikiton-N@Mail.Ru>

N-Nagorny commented on docs/Behaviour - Server Side.md:
```diff
@@ -21,9 +21,9 @@ A Sender managed with IS-11 has the following states:
-- `non_compliant_flow` when this Receiver is active and the Flow is non compliant with Receiver Capabilities of this Receiver (e.g. there is an active `transport_file` which violates the Receiver Capabilities). If the Receiver becomes inactive this state is preserved until the next activation.
+- `unknown` when it is not possible to say whether the active Flow is compliant with Receiver Capabilities of this Receiver (e.g. it is activated without a `transport_file`).
+- `compliant_flow` when the active Flow is compliant with Receiver Capabilities of this Receiver (e.g. there is an active `transport_file` and it does not violate the Receiver Capabilities).
+- `non_compliant_flow` when the active Flow is non compliant with Receiver Capabilities of this Receiver (e.g. there is an active `transport_file` which violates the Receiver Capabilities). If the Receiver becomes inactive this state is preserved until the next activation.
```
```suggestion
- `unknown` when it is not possible to say whether the active stream is compliant with Receiver Capabilities of this Receiver (e.g. it is activated without a `transport_file`).
- `compliant_stream` when the active stream is compliant with Receiver Capabilities of this Receiver (e.g. there is an active `transport_file` and it does not violate the Receiver Capabilities).
- `non_compliant_stream` when the active stream is non compliant with Receiver Capabilities of this Receiver (e.g. there is an active `transport_file` which violates the Receiver Capabilities). If the Receiver becomes inactive this state is preserved until the next activation.
```

N-Nagorny replied:
Just to not forget: `non_compliant_flow` -> `non_compliant_stream`

alabou committed:
Update docs/Behaviour - Client Side.md from Nikita comments

Co-authored-by: N-Nagorny <Nikiton-N@Mail.Ru>

alabou committed:
Update docs/Behaviour - Server Side.md from Nikita comments

Co-authored-by: N-Nagorny <Nikiton-N@Mail.Ru>

alabou committed:
Update Behaviour - Server Side.md

as per meeting receiver state *_flow -> *_stream

alabou committed:
Update Behaviour - Client Side.md

As per the meeting discussions ... replacing Nikita wording slightly from "If the processing Constraint Sets results in an empty constraint_sets of Active Constraints, NMOS Controller SHOULD/MAY inform the user about it." to "If the processing of the Receiver Capabilities results in an empty `constraint_sets` of Active Constraints, NMOS Controller SHOULD inform the user about it."

alabou committed:
Update receiver-status.json

as per meeting decision *_flow -> *_stream

alabou committed:
Update receiver-status-get-200.json

as per meeting decision *_flow -> *_stream

alabou committed:
Update FlowCompatibilityManagementAPI.raml

as per meeting discussions *_flow -> *_stream

alabou commented:
I included Nikita's comments/suggestions and some other edits as discussed in our meeting of March 21th in a few additional commits. Make sure to compare using all the commits of alabou-patch-1 from v1.0-dev.

N-Nagorny commented on docs/Behaviour - Client Side.md:
```diff
@@ -2,22 +2,22 @@
+`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Receiver Capabilities indicates relative preference between Constraint Sets of that Receiver. The NMOS Controller MAY use these values to make a decision about how to process these Constraint Sets and determine preference of Constraint Sets of Active Constraints. It SHOULD NOT propagate values of this property as is from Receiver Cap ...
 
-`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Active Constraints indicates relative preference between Constraints Sets of the Active Constraints of the Sender. The NMOS Controller MAY fill in this property in Constraint Sets based on any additional information about Receivers. For example, if the user in any way informs the NMOS Controller that Receiver D is more important than ...
+`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Active Constraints indicates relative preference between Constraint Sets of the Active Constraints of the Sender. The NMOS Controller MAY fill in this property in Constraint Sets based on any additional information about Receivers. For example, if the user in any way informs the NMOS Controller that Receiver D preferred Constraint Se ...
```
Shouldn't it state the following?
```suggestion
`urn:x-nmos:cap:meta:preference` property in Constraint Sets of Active Constraints indicates relative preference between Constraint Sets of the Active Constraints of the Sender. The NMOS Controller MAY fill in this property in Constraint Sets based on any additional information about Receivers. For example, if the user in any way informs the NMOS Controller that Receiver D has precedence over the consensus of Receivers A, B, C and `urn:x-nmos:cap:meta:preference` values of Receiver D state that Constraint Set 6 has higher preference than Constraints Sets 2, 3, 4, 5, then the Constraint Sets would be 2, 3, 4, 5, 6 where Constraint Set 6 has the highest preference.
```

alabou replied:
I think that it is not only Receiver D that has a preference from the user but one constraint set of such Receiver D ...

alabou committed:
Update docs/Behaviour - Client Side.md (from Nikita comment)

Co-authored-by: N-Nagorny <Nikiton-N@Mail.Ru>
